ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Node/GSM.hs

```diff
@@ -0,0 +1,327 @@
+  deriving (Eq, Show)
+
+data GsmView m blk upstreamPeer selection candidate = GsmView {
+    candidateOverSelection    :: selection -> candidate -> CandidateVersusSelection
+  ,
+    durationUntilTooOld       :: Maybe (selection -> m DurationFromNow)
+    -- ^ How long from now until the selection will be so old that the node
+    -- should exit the @CaughtUp@ state
+    --
+    -- 'Nothing' means the selection can never become too old.
+  ,
+    equivalent                :: selection -> selection -> Bool
+    -- ^ Whether the two selections are equivalent for the purpose of the
+    -- Genesis State Machine
+  ,
+    getChainSyncCandidates    :: STM m (Map.Map upstreamPeer (StrictSTM.StrictTVar m candidate))
+    -- ^ The latest candidates from the upstream ChainSync peers
+  ,
+    getChainSyncIdlers       :: STM m (Set.Set upstreamPeer)
+    -- ^ The ChainSync peers whose latest message claimed that they have no
+    -- subsequent headers
+  ,
+    getCurrentSelection       :: STM m selection
+    -- ^ The node's current selection
+  ,
+    setCaughtUpPersistentMark :: Bool -> m ()
+    -- ^ EG touch/delete the marker file on disk
+  ,
+    varLedgerStateJudgement   :: LazySTM.TVar m LedgerStateJudgement
+    -- ^ EG update the TVar that the diffusion layer monitors
+  }
+
+-- | The two proper GSM states for boot strap peers
+--
+-- See the @BootstrapPeersIER.md@ document for their specification.
+data GsmEntryPoints m = GsmEntryPoints {
+    enterCaughtUp      :: forall neverTerminates. m neverTerminates
+  ,
+    enterOnlyBootstrap :: forall neverTerminates. m neverTerminates
+  }
+
+{-------------------------------------------------------------------------------
+  A real implementation
+-------------------------------------------------------------------------------}
+
+-- | The actual GSM logic for boot strap peers
+--
+-- See the @BootstrapPeersIER.md@ document for the specification of this logic.
+realGsmEntryPoints :: forall m blk upstreamPeer selection candidate.
+     ( MonadTimer m
+     , Eq upstreamPeer
+     )
+  => GsmView m blk upstreamPeer selection candidate
+  -> GsmEntryPoints m
+realGsmEntryPoints gsmView = GsmEntryPoints {
+    enterCaughtUp
+  ,
+    enterOnlyBootstrap
+  }
+  where
+    GsmView {
+        candidateOverSelection
+      ,
+        durationUntilTooOld
+      ,
+        equivalent
+      ,
+        getChainSyncCandidates
+      ,
+        getChainSyncIdlers
+      ,
+        getCurrentSelection
+      ,
+        setCaughtUpPersistentMark
+      ,
+        varLedgerStateJudgement
+      } = gsmView
+
+    enterCaughtUp :: forall neverTerminates. m neverTerminates
+    enterCaughtUp = do
+        atomically $ LazySTM.writeTVar varLedgerStateJudgement YoungEnough
+        setCaughtUpPersistentMark True
+
+        blockWhileCaughtUp
+        enterOnlyBootstrap
+
+    enterOnlyBootstrap :: forall neverTerminates. m neverTerminates
+    enterOnlyBootstrap = do
+        setCaughtUpPersistentMark False
```

Comment:
This line should instead be in `enterCaughtUp`, just before the call to `enterOnlyBootstrap`. Locally, this is indistinguishable. However, the node start up calls `enterOnlyBootstrap` exactly when the file is not present (ie when it's _already_ `False`).

cc @bolt12

Reply:
I pushed up what I think is a fix for this.